bin/flatcar-install: run write_to_disk from subshell

tormath1 opened the pull request:
Somehow, with the bash upgrade the trap RETURN signal is not handled correctly (it's ignored).

Adding an explicit return makes it work.

<hr>

The bash upgrade and the way it's done (a single huge commit) it's too complicated to make a proper bisect but I assume the issue is regarding the changes on how trap signals are handled.

## Testing done
* Locally
* Implementing Mantle test for this script

pothos commented on bin/flatcar-install:
```diff
@@ -596,7 +596,7 @@ function write_to_disk() {
     done
     [ -z "$try" ] || exit 1
-    udevadm settle
+    udevadm settle && return
```
Would this work, too, because when udevadm settle fails, the script exists due to  `set -e`?
```suggestion
    udevadm settle
    return
```

pothos replied:
A comment for this workaround would also be good

tormath1 replied:
Thanks for the suggestion - it works fine.

pothos replied:
Why do we need the trap? Can't the write the "done" in the last line instead of the `return`?

pothos replied:
Ok, I think the idea was to write the "done" as soon as the function did something, instead of the last line it should be done as first.

pothos replied:
The current approach with the explicit `return` doesn't cover the case where `dd` failed and thus the `is_modified` check wouldn't work anymore.

tormath1 replied:
I followed @krnowak approach and use a subshell + EXIT trap instead of explicit `return` - but I still can't explain why it works on Equinix Metal setup.

krnowak commented:
I remember replacing all RETURN traps to EXIT traps in subshells because of that. See it here: https://github.com/flatcar/scripts/commit/eee6b50aa57fae7e6b12f1da3ef7356d335a20c8

pothos commented:
The install script is tested in the Equinix Metal installation path, strange that we didn't see the issue there?

pothos commented on bin/flatcar-install:
```diff
@@ -577,8 +577,11 @@ function wait_for_disk() {
+    (
     mkfifo -m 0600 "${WORKDIR}/disk_modified"
-    trap '(exec 2>/dev/null ; echo done > "${WORKDIR}/disk_modified") &' RETURN
+    trap '(exec 2>/dev/null ; echo done > "${WORKDIR}/disk_modified") &' EXIT
```
```suggestion
    (exec 2>/dev/null ; echo done > "${WORKDIR}/disk_modified") &
```
I'm fine with the current approach, my idea was to simplify this and don't use traps.
This should work because we don't wait for the write somewhere in parallel but it's all sequential code and the marker can be written early.

tormath1 replied:
Removing the trap is creating a weird behavior:
```
++ write_to_disk
++ mkfifo -m 0600 /tmp/flatcar-install.r3mrgIyVoA/disk_modified
+ wget --tries 10 --timeout=20 --retry-connrefused --no-verbose -O - https://beta.release.flatcar-linux.net/amd64-usr/3602.1.0/flatcar_production_image.bin.bz2
+ gpg --batch --trusted-key E25D9AED0593B34A --verify /tmp/flatcar-install.r3mrgIyVoA/flatcar_production_image.bin.bz2.sig -
+ tee /dev/fd/62
++ lbzip2 -cd
+++ blockdev --getsz /dev/sda
++ exec
++ dd conv=nocreat count=1024 if=/dev/zero of=/dev/sda seek=41942016 status=none
++ dd bs=1M conv=nocreat of=/dev/sda status=none
2023-06-23 09:40:39 URL:https://beta.release.flatcar-linux.net/amd64-usr/3602.1.0/flatcar_production_image.bin.bz2 [379416813/379416813] -> "-" [1]
gpg: Signature made Mon May 29 09:40:52 2023 UTC
gpg:                using RSA key 8D6DA7853CFE1B1ED346ED0DEDBDF411267EC954
gpg:                issuer "buildbot@flatcar-linux.org"
gpg: key E25D9AED0593B34A marked as ultimately trusted
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
gpg: Good signature from "Flatcar Buildbot (Official Builds) <buildbot@flatcar-linux.org>" [ultimate]
+ wait_for_disk
+ '[' -n '' ']'
+ read -rt 7200 _disk_status
+ write_cloudinit
+ [[ -n '' ]]
+ write_ignition
+ [[ -n '' ]]
+ [[ -n '' ]]
+ echo 'Success! Flatcar Container Linux beta 3602.1.0 is installed on /dev/sda'
Success! Flatcar Container Linux beta 3602.1.0 is installed on /dev/sda
+ rm -rf /tmp/flatcar-install.r3mrgIyVoA
+ trap - EXIT
```

It ends right after the `dd` (it does not go through the `udevadm settle` and the for loop) so it takes a few second before the changes being visible

pothos replied:
Ah, `install_from_url` doesn't wait for `write_to_disk` to finish. It even doesn't capture the error code because  `> >(write_to_disk)` redirects the output to the input FD but it's not a pipe where `pipefail` would work. This should be fixed to properly fail on error.

pothos replied:
Prepared a fix in https://github.com/flatcar/init/pull/99

pothos replied:
Had to touch the waiting logic anyway now to signal the failure and removed the trap.

pothos reviewed (commented):
There are two other `RETURN` traps for `umount`, do you want to use the subshell trick there?

tormath1 committed:
bin/flatcar-install: use EXIT instead of RETURN traps

Somehow, with the bash upgrade the trap RETURN signal is not handled
correctly (it's ignored).

We run now everything from a subshell and turn RETURN trap to an EXIT
trap.

Signed-off-by: Mathieu Tortuyaux <mtortuyaux@microsoft.com>
